### PR TITLE
parser, ddl: data type default values support `UPPER(SUBSTRING_INDEX(user(),'@',1))`

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/50966, close https://github.com/pingcap/tidb/issues/51200

Problem Summary:

### What changed and how does it work?
* Data type default values support `UPPER(SUBSTRING_INDEX(user(),'@',1))`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @tangenta @CbcWestwolf

File: pkg/ddl/ddl_api.go [hunk 1]
```diff
@@ -1339,6 +1339,32 @@ func getFuncCallDefaultValue(col *table.Column, option *ast.ColumnOption, expr *
...
+		if substringIndexFunc, ok := expr.Args[0].(*ast.FuncCallExpr); ok && substringIndexFunc.FnName.L == ast.SubstringIndex {
+			if err := expression.VerifyArgsWrapper(substringIndexFunc.FnName.L, len(substringIndexFunc.Args)); err != nil {
+				return nil, false, errors.Trace(err)
+			}
+			if userFunc, ok := substringIndexFunc.Args[0].(*ast.FuncCallExpr); ok && userFunc.FnName.L == ast.User {
+				if err := expression.VerifyArgsWrapper(userFunc.FnName.L, len(userFunc.Args)); err != nil {
+					return nil, false, errors.Trace(err)
+				}
+				valExpr, isValue := substringIndexFunc.Args[1].(ast.ValueExpr)
+				if !isValue || valExpr.GetString() != "@" {
```
Comment (tangenta) [inline pkg/ddl/ddl_api.go:1356]: Why not support a wider scope like `upper(substring_index(*))` or `upper(*)`?

Comment (zimulala) [reply pkg/ddl/ddl_api.go:1356]: For the time being, only the expressions that need to be supported are supported. In addition, these default expressions are supported by different versions of mysql (5.7, 8.0, 8.0.34, etc.).

**Review by tangenta — APPROVED** [review]

**Review by CbcWestwolf — APPROVED** [review]

**Pull request merged**